contracts/src/Gateway.sol

```diff
@@ -493,199 +391,182 @@ contract Gateway is IGateway, IInitializable, IUpgradable {
+    /// @dev Submit a message from Polkadot for verification and dispatch
+    /// @param message A message produced by the OutboundQueue pallet on BridgeHub
+    /// @param leafProof A message proof used to verify that the message is in the merkle tree committed by the OutboundQueue pallet
+    /// @param headerProof A proof that the commitment is included in parachain header that was finalized by BEEFY.
+    /// @param rewardAddress Account on BH to credit delivery rewards
+    function v2_submit(
+        InboundMessageV2 calldata message,
+        bytes32[] calldata leafProof,
+        Verification.Proof calldata headerProof,
+        bytes32 rewardAddress
+    ) external {
+        CoreStorage.Layout storage $ = CoreStorage.layout();
 
-        // Destination fee always in DOT
-        uint256 fee = _calculateFee(ticket.costs);
+        bytes32 leafHash = keccak256(abi.encode(message));
 
-        // Ensure the user has enough funds for this message to be accepted
-        if (msg.value < fee) {
-            revert FeePaymentToLow();
+        if ($.inboundNonce.get(message.nonce)) {
+            revert IGateway.InvalidNonce();
         }
 
-        channel.outboundNonce = channel.outboundNonce + 1;
+        $.inboundNonce.set(message.nonce);
 
-        // Deposit total fee into agent's contract
-        payable(channel.agent).safeNativeTransfer(fee);
+        // Produce the commitment (message root) by applying the leaf proof to the message leaf
+        bytes32 commitment = MerkleProof.processProof(leafProof, leafHash);
 
-        // Reimburse excess fee payment
-        if (msg.value > fee) {
-            payable(msg.sender).safeNativeTransfer(msg.value - fee);
+        // Verify that the commitment is included in a parachain header finalized by BEEFY.
+        if (!_verifyCommitment(commitment, headerProof)) {
+            revert IGateway.InvalidProof();
         }
 
-        // Generate a unique ID for this message
-        bytes32 messageID = keccak256(abi.encodePacked(channelID, channel.outboundNonce));
+        bool success = v2_dispatch(message);
 
-        emit IGateway.OutboundMessageAccepted(channelID, channel.outboundNonce, messageID, ticket.payload);
+        emit IGateway.InboundMessageDispatched(message.nonce, success, rewardAddress);
     }
 
-    /// @dev Outbound message can be disabled globally or on a per-channel basis.
-    function _ensureOutboundMessagingEnabled(Channel storage ch) internal view {
-        CoreStorage.Layout storage $ = CoreStorage.layout();
-        if ($.mode != OperatingMode.Normal || ch.mode != OperatingMode.Normal) {
-            revert Disabled();
+    function v2_dispatch(InboundMessageV2 calldata message) internal returns (bool) {
+        for (uint256 i = 0; i < message.commands.length; i++) {
+            if (gasleft() * 63 / 64 < message.commands[i].gas + DISPATCH_OVERHEAD_GAS_V2)
+            {
+                assembly {
+                    invalid()
+                }
+            }
```

Comment:
Is this check required for V2?

Reply:
Yes it is. It there to ensure that relayers pass enough gas to allow each command in the msg to dispatch.

Reply:
So we need to ensure that the outbound-queue pallet continues to compute the maximum dispatch gas for each command in the msg.